data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/ScanObjectWorker.java

```diff
@@ -0,0 +1,184 @@
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+/**
+ * Class responsible for processing the s3 scan objects with the help of <code>S3ObjectWorker</code>
+ * or <code>S3SelectWorker</code>.
+ */
+public class ScanObjectWorker implements Runnable{
+
+    private static final Logger LOG = LoggerFactory.getLogger(ScanObjectWorker.class);
+    private final S3Client s3Client;
+    private final S3AsyncClient s3AsyncClient;
+    static final char DAYS = 'd';
+    static final char WEEKS = 'w';
+    static final char MONTHS = 'm';
+    static final char YEARS = 'y';
+    private final S3ObjectPluginMetrics s3ObjectPluginMetrics;
+    private final Buffer<Record<Event>> buffer;
+    private final int numberOfRecordsToAccumulate;
+    private final Duration bufferTimeout;
+    private final BucketOwnerProvider bucketOwnerProvider;
+    private final BiConsumer<Event, S3ObjectReference> eventMetadataModifier;
+    private static final Map<String,S3ObjectDetails> stateSaveMap = new HashMap<>();
+    private final List<ScanOptions> scanOptionsBuilderList;
+
+    public ScanObjectWorker(final S3ObjectRequest s3ObjectRequest,final List<ScanOptions> scanOptionsBuilderList){
+        this.s3Client = s3ObjectRequest.getS3Client();
+        this.s3AsyncClient = s3ObjectRequest.getS3AsyncClient();
+        this.scanOptionsBuilderList = scanOptionsBuilderList;
+        this.buffer = s3ObjectRequest.getBuffer();
+        this.s3ObjectPluginMetrics = s3ObjectRequest.getS3ObjectPluginMetrics();
+        this.numberOfRecordsToAccumulate = s3ObjectRequest.getNumberOfRecordsToAccumulate();
+        this.bufferTimeout = s3ObjectRequest.getBufferTimeout();
+        this.eventMetadataModifier = s3ObjectRequest.getEventConsumer();
+        this.bucketOwnerProvider = s3ObjectRequest.getBucketOwnerProvider();
+    }
+
+    /**
+     * It will decide the s3 object parse <code>S3ObjectWorker</code> or <code>S3SelectWorker</code>
+     * based on s3 select configuration provided.
+     */
+    @Override
+    public void run() {
+        scanOptionsBuilderList.forEach(this::parseS3ScanObjects);
+    }
+
+    /**
+     * Method will parse the s3 object and write to {@link Buffer}
+     */
+    void parseS3ScanObjects(final ScanOptions scanOptionsBuilder) {
+        final S3ObjectRequest.Builder s3ObjectRequestBuilder = new S3ObjectRequest.Builder(buffer, numberOfRecordsToAccumulate,
+                bufferTimeout, s3ObjectPluginMetrics);
+        final S3ObjectHandler s3ObjectHandler;
+        if(scanOptionsBuilder.getExpression() != null){
```

Comment:
Most of this code is duplicated with code in `S3Source`. This class should use the same `S3ObjectHandler` as the SQSWorker uses. This can be injected into this class.

https://github.com/opensearch-project/data-prepper/blob/068c57172c0dfb0c4e43d1d796e8a59fa11fb1df/data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/S3Source.java#L72-L98